src/ShopPaymentTransactionBundle/objects/TPkgShopPaymentTransactionManagerEndPoint.class.php

```diff
@@ -223,6 +225,74 @@ protected function updateOrderPaidStatusBasedOnCurrentBalance()
         return $bOrderPaidStatusChanged;
     }
 
+    private function updateRealUsedVoucherValueBasedOnTransactions(TdbPkgShopPaymentTransaction $transaction): void
+    {
+        $oTransactionData = $this->getTransactionDataFromOrder(
+            TPkgShopPaymentTransactionData::TYPE_PAYMENT,
+            null,
+            true
+        );
+        $valueToCapture = $oTransactionData->getTotalValueForItemType(TPkgShopPaymentTransactionItemData::TYPE_PRODUCT);
+
+        // There are uncaptured products, so no voucher handling is needed because not captured products can use open/refunded voucher values.
+        if ($valueToCapture > 0) {
+            return;
+        }
+
+        $realVoucherValueUsed = $this->getTransactionPositionTotalForType(
+                TPkgShopPaymentTransactionItemData::TYPE_VOUCHER,
+                true
+            ) * -1;
+        $shopOrder = $transaction->GetFieldShopOrder();
+        if (null === $shopOrder) {
+            return;
+        }
+
+        $this->updateRealUsedVoucherValue($shopOrder, $realVoucherValueUsed);
+    }
+
+    private function updateRealUsedVoucherValue(TdbShopOrder $shopOrder, float $realVoucherValueUsed): void
+    {
+        $voucherUsedDifference = $shopOrder->fieldValueVouchers - $realVoucherValueUsed;
+        if ($voucherUsedDifference > 0) {
```

Comment:
Please try to avoid deeply nested methods, e.g. by using early returns (and by splitting methods even more).

Reply:
done